src/centrality.rs

```diff
@@ -0,0 +1,217 @@
+        for i in 0..betweenness.len() as usize {
+            betweenness[i] *= scale;
+        }
+    }
+}
+
+fn _accumulate_basic(
+    betweenness: &mut Vec<f64>,
+    node_count: usize,
+    sorted_by_distance: &Vec<NodeIndex>,
+    mut predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+    is: usize,
+) {
+    let mut delta = vec![0.0; node_count];
+    for w in sorted_by_distance {
+        let iw = (*w).index();
+        let coeff = (1.0 + delta[iw]) / (sigma[&iw] as f64);
+        let p_w = predecessors.get_mut(&w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] += delta[iw];
+        }
+    }
+}
+
+fn _accumulate_endpoints(
+    betweenness: &mut Vec<f64>,
+    node_count: usize,
+    sorted_by_distance: &Vec<NodeIndex>,
+    mut predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+    is: usize,
+) {
+    betweenness[is] += (sorted_by_distance.len() - 1) as f64;
+    let mut delta = vec![0.0; node_count];
+    for w in sorted_by_distance {
+        let iw = (*w).index();
+        let coeff = (1.0 + delta[iw]) / (sigma[&iw] as f64);
+        let p_w = predecessors.get_mut(&w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] += delta[iw] + 1.0;
+        }
+    }
+}
+
+fn shortest_path_for_centrality<G>(
+    graph: G,
+    node_s: &G::NodeId,
+) -> (
+    Vec<NodeIndex>,
+    HashMap<G::NodeId, Vec<G::NodeId>>,
+    HashMap<usize, i64>,
+)
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + GraphBase<NodeId = NodeIndex>, // for get() and get_mut()
+{
+    let mut sorted_by_distance: Vec<NodeIndex> = Vec::new(); // a stack
+    #[allow(non_snake_case)]
+    let mut Q: Vec<NodeIndex> = Vec::new(); // a queue
+    let mut predecessors = HashMap::<G::NodeId, Vec<G::NodeId>>::new();
+    let mut sigma = HashMap::<usize, i64>::new();
+    let mut distance = HashMap::<G::NodeId, i64>::new();
+
+    let i_s = graph.to_index(*node_s);
+    let index_s = NodeIndex::new(i_s);
+
+    for node in graph.node_identifiers() {
+        let i = graph.to_index(node);
+        let index = NodeIndex::new(i);
+        predecessors.insert(index, Vec::new());
+        sigma.insert(i, 0);
+        distance.insert(index, -1);
+    }
+    sigma.insert(index_s.index(), 1);
+    distance.insert(index_s, 0);
+    Q.push(index_s);
+    while !Q.is_empty() {
+        let v = Q.remove(0);
+        sorted_by_distance.push(v);
+        //        let neighbors = graph.neighbors_directed(v, Incoming);
+        for w in graph.neighbors(v) {
+            if *(distance.get(&w).unwrap()) < 0 {
+                Q.push(w);
+                distance.insert(w, *(distance.get(&v).unwrap()) + 1);
```

Comment:
One thing I've been using more is `distance[&v]` in cases like this where we know v is in `distance` just to avoid the need to unwrap since it fails if v isn't in the hashmap instead of returning `Option<T>`

Reply:
OMG. I just successfully replaced `*(distance.get(&w).unwrap())` with `distance[&w]`. I am now more positively disposed towards rust.

Reply:
Is there a simpler way to write `distance.insert(w, distance[&v] + 1)`? I'd like to write `distance[&w] = distance[&v] + 1`.

Reply:
I don't think HashMap implements the [`IndexMut` trait](https://doc.rust-lang.org/std/ops/trait.IndexMut.html) which allows mutating like that. So the only way I can see it working is if you know that `distance[w]` is always defined you can do something like:

```rust
*distance.get_mut(&w).unwrap() = distance[&v] + 1
```

If it's not guaranteed to be defined (so `unwrap()` could panic) you use an entry. so do something like:

```rust
distance.entry(&w).and_modify(|_| {distance[&v] + 1}).or_insert(distance[&v] + 1);
```

(I guess you can use the entry to do `distance.entry(&w).and_modify(|_| {distance[&v] + 1});` too to avoid the `get_mut` and `unwrap()`)

Reply:
This is probably resolved with recent edits.